tests/common/graceful_shutdown_test.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build e2e
+
+package common
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/interfaces"
+)
+
+func TestGracefulShutdown(t *testing.T) {
+	testRunner.BeforeTest(t)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	clus := testRunner.NewCluster(ctx, t, config.WithClusterSize(3))
+	// clean up orphaned resources like closing member client.
+	defer clus.Close()
+	// shutdown each etcd member process sequentially
+	// and start from old leader, (new leader), (follower)
+	members := clus.Members()
+	mn := len(members)
+	for len(members) > 0 {
+		// transfer leadership only when quorum is active after leader shuts down
+		expectTransferLeadership := len(members)-1 >= 1+(mn-1)/2
+		members = tryShutdownLeader(ctx, t, members, expectTransferLeadership)
+	}
+}
+
+// tryShutdownLeader tries stop etcd member if it is leader. If no leader is found, stop the very first member.
+// it also asserts stop member should not take longer than 1.5 seconds and lead ID has been changed.
+func tryShutdownLeader(ctx context.Context, t *testing.T, members []interfaces.Member, expectTransferLeadership bool) (remainingMembers []interfaces.Member) {
```

Comment:
This function combines test execution and validation. Can you instead extract function to get leader and have TestGracefulShutdown do:
```
    
  	for len(members)-1 >= 1+(mn-1)/2 {
        leader, followers := getLeader(members)
        
        start := time.Now()
        leader.Stop()
        require.Less(t, time.Since(start), 1500*time.Millisecond)
        
        resps, err := followers[0].Client().Status(ctx)
		require.NoError(t, err)
	    require.NotEqualf(t, resps[0].Leader, leader.ID(), "expect old leaderID %x changed to new leader ID %x", leader.ID(), resps[0].Leader)
        
		members = followers
	}
```